Implement inserts

krishnangovindraj opened the pull request.

krishnangovindraj committed:
STASH

krishnangovindraj committed:
WIP: Done Insert Planning

krishnangovindraj committed:
WIP: Write some tests

krishnangovindraj committed:
Fix fallout from rebase

krishnangovindraj committed:
WIP: Oh boy, I have stash problems

krishnangovindraj committed:
Does that work?

krishnangovindraj committed:
Add assert for the test

krishnangovindraj committed:
Add insert test, find problem with roles, be sad.

krishnangovindraj committed:
Flip how types are indexed; Introduce type_reader method to get roles by name

krishnangovindraj committed:
Insert test for has and roleplayers run

krishnangovindraj committed:
Bump typedb behaviour

krishnangovindraj committed:
Get some BDD running

krishnangovindraj committed:
Fix fallout from rebase

krishnangovindraj commented on compiler/inference/type_seeder.rs:
```diff
@@ -491,6 +503,26 @@ impl UnaryConstraint for Label<Variable> {
+        }
+    }
+}
+
```
RoleName is handled as a `UnaryConstraint`, analogous to `Label`

krishnangovindraj commented on concept/type_/type_manager/type_cache/type_cache.rs:
```diff
@@ -121,6 +122,14 @@ impl TypeCache {
+            }
+            role_types_by_name.get_mut(label.name.as_str()).unwrap().push(role_type.clone());
+        }
+
```
Custom logic to build the cache. This is easier than using `TypeReader`.

krishnangovindraj commented on concept/type_/type_manager/type_reader.rs:
```diff
@@ -56,7 +57,8 @@ impl TypeReader {
     {
-        let key = LabelToTypeVertexIndex::build(label.scoped_name.as_reference()).into_storage_key();
+        let key =
+            LabelToTypeVertexIndex::build(label.inverted_scoped_name_for_index().as_reference()).into_storage_key();
```
The index now stores `Prefix + name [ + ':' + scope ]` instead of `Prefix [ + ':' + scope ]  + name`
This allows us to query roles by name as well as scoped name. The original would let us to query by scope and scoped name, which was less useful.

flyingsilverfin replied:
good idea

krishnangovindraj commented on encoding/value/label.rs:
```diff
@@ -83,6 +83,18 @@ impl<'a> Label<'a> {
+            StringBytes::new(Bytes::reference(self.name.bytes().bytes()))
+        }
+    }
+
```
This may be better on the trait used by the index.

krishnangovindraj commented on ir/pattern/constraint.rs:
```diff
@@ -92,6 +92,18 @@ impl<'cx> ConstraintsBuilder<'cx> {
+        let as_ref = self.constraints.add_constraint(role_name);
+        Ok(as_ref.as_role_name().unwrap())
+    }
+
```
New constraint, because it's not the same as using a `Label` (which loses the context).

krishnangovindraj commented on ir/translation/constraints.rs:
```diff
@@ -165,16 +184,22 @@ fn add_typeql_isa(
+            let assigned = extend_from_inline_typeql_expression(function_index, constraints, expression)?;
+            let attribute = constraints.create_anonymous_variable()?;
+            constraints.add_comparison(attribute, assigned)?; // TODO: I should probably not piggy back on comparison like this.
+            attribute
```
This was copying what we did in `2.x` but maybe we want an explicit constraint for cases like this.

flyingsilverfin replied:
An equality is probs ok

krishnangovindraj commented on ir/translation/constraints.rs:
```diff
@@ -190,9 +215,26 @@ fn add_typeql_relation(
+                    TypeRefAny::Type(TypeRef::Named(NamedType::BuiltinValueType(_))) => {
+                        unreachable!("Why do we allow value types here?")
+                    }
+                };
```
Why not make the parsed be more specific?

flyingsilverfin replied:
Not sure what you mean, but this probably works to unblock inserts, please make a list of notes for yourself we can review together in person on Monday, including this!

krishnangovindraj commented on ir/translation/insert.rs:
```diff
@@ -0,0 +1,25 @@
+) -> Result<FunctionalBlockBuilder, PatternDefinitionError> {
+    let mut builder = FunctionalBlock::builder();
+    for statement in &insert.statements {
+        add_statement(function_index, &mut builder.conjunction_mut().constraints_mut(), statement)?;
```
By re-using the builder, it parses everything it would in a match and fails when compiling it to the executor.

krishnangovindraj commented on executor/insert_executor.rs:
```diff
@@ -0,0 +1,199 @@
+
+pub struct InsertExecutor {
+    plan: InsertPlan,
+    reused_created_things: Vec<answer::Thing<'static>>, // internal mutability is cleaner
```
Re-used vector for pushing created concepts. The order follows that of instructions in the insertion plan

krishnangovindraj commented on compiler/planner/insert_planner.rs:
```diff
@@ -0,0 +1,402 @@
+        .index
+        .iter()
+        .map(|(_, i)| VariableSource::ThingSource(ThingSource::Inserted(*i)))
+        .collect::<Vec<_>>();
```
Should the output row be planned? Or should we just return `concat(copy(input_row), inserted)`

flyingsilverfin replied:
output row ordering doesn't really matter, no

krishnangovindraj commented on compiler/planner/insert_planner.rs:
```diff
@@ -0,0 +1,402 @@
+        };
+        instructions.push(instruction);
+        Ok(())
+    })?;
```
`isa` are executed first so all vertices are available before we move on to the edges.
Even concepts created using `has <type> <value>` will have an `isa` statement.

krishnangovindraj commented on compiler/planner/insert_planner.rs:
```diff
@@ -0,0 +1,402 @@
+
+#[derive(Debug)]
+pub enum InsertInstruction {
+    // TODO: Just replace this with regular `Constraint`s and use a mapped-row?
+    Entity { type_: TypeSource },
+    Attribute { type_: TypeSource, value: ValueSource },
+    Relation { type_: TypeSource },
+    Has { owner: ThingSource, attribute: ThingSource }, // TODO: Ordering
+    RolePlayer { relation: ThingSource, player: ThingSource, role: TypeSource }, // TODO: Ordering
```
This feels decoupled and easier to reason about than using the constraints directly.

krishnangovindraj commented on compiler/planner/insert_planner.rs:
```diff
@@ -0,0 +1,402 @@
+    TypeSource(TypeSource),
+    ValueSource(ValueSource),
+    ThingSource(ThingSource),
+}
```
Similar purpose as `VariablePosition`, but wrapped to indicate which array to look in.

krishnangovindraj committed:
Self review cleanup

krishnangovindraj committed:
WIP: Think about pipeline stuff

flyingsilverfin commented on answer/variable_value.rs:
```diff
@@ -32,6 +32,14 @@ impl<'a> VariableValue<'a> {
+    pub fn as_value(&self) -> &Value<'a> {
+        match self {
+            VariableValue::Value(value) => value,
+            // TODO: Do we want to implicit cast from attributes?
```
My general thought was that at the implementation we can keep explicitly Values as Value variable contents, and Attributes in Thing variables, but have Attributes be usable as Values with only explicit conversion. However, this is really hard for me to judge without playing with the code. Would you be able to avoid resolving this until next week when I'm back? I think is quite important!

krishnangovindraj replied:
I agree, I think we should ban it and add casts into the IR.

flyingsilverfin commented on compiler/planner/insert_planner.rs:
```diff
@@ -0,0 +1,402 @@
+    let value_constants = collect_values_from_expression_and_comparison(constraints, &input_variables)?;
+    let isa_kinds = collect_kind_from_isa(constraints, input_variables, type_annotations)?;
+
+    let mut instructions = Vec::new();
```
maybe can pre-size this (even approximately)?

flyingsilverfin commented on concept/type_/type_manager/type_writer.rs:
```diff
@@ -69,7 +69,7 @@ impl<Snapshot: WritableSnapshot> TypeWriter<Snapshot> {
         Self::storage_put_type_vertex_property(snapshot, type_.clone(), Some(label.clone().into_owned()));
 
-        let label_to_vertex_key = LabelToTypeVertexIndex::build(label.scoped_name());
+        let label_to_vertex_key = LabelToTypeVertexIndex::build(label.inverted_scoped_name_for_index());
```
rather than Label having the method to order themselves for the index key builder, can the LabelToTypeVertexIndex do this inside the `build` method and take a Label? Label  lives in encoding so IMO it should work. This way it currently leaks indexing implementation into the Label data structure, which is non-ideal

flyingsilverfin replied:
Do we already have a pattern where the Index/Vertex etc takes an object and builds a key for that internally? The representation is chosen by the internal implementation of the Index/Vertex encoding

flyingsilverfin commented on executor/insert_executor.rs:
```diff
@@ -0,0 +1,193 @@
+        value_constants: &'a Vec<Value<'static>>,
+        created_things: &'a mut Vec<answer::Thing<'static>>,
+    ) -> ExecutionConcepts<'a, 'row> {
+        // TODO: Should we keep an attribute cache to avoid re-creating attributes with the same value?
```
unsure, skip unless obviously better

krishnangovindraj committed:
Move to traits implementing the actual insertion logic

krishnangovindraj committed:
Partially implement define queries (#7126)

## Usage and product changes
Implement basic functionality of `define` queries.

## Implementation
We iterate over the parsed `definable`s and apply changes in the following order:
`types ≤ type annotations < sub < value-type < relates < plays ≤ owns`

Some functionality remains:
* Implement overrides,
* Figure out how to elegantly handle idempotence and conflicting declarations in a define query. e.g.
```
define
relation rel, relate rol[];
relation subrel relates subrol as rol;
```

krishnangovindraj committed:
WIP: Refactor to simplify things

krishnangovindraj committed:
Stash to rebase

krishnangovindraj committed:
Fix rebase breaks

krishnangovindraj committed:
Delete test works?

dmitrii-ubskii committed:
Include type annotations into compiled instructions (#7131)

krishnangovindraj committed:
Introduce IR for schema constraints (#7129)

## Usage and product changes
Introduce `Owns`, `Relates` & `Plays` constraints into IR &  implement type-inference traits for them.

## Implementation
Introduce `Owns`, `Relates` & `Plays` constraints into IR &  implement type-inference traits for them.

krishnangovindraj committed:
Cleanup post rebase

krishnangovindraj committed:
Clean up commented out code

krishnangovindraj committed:
Cleanup, fix build

krishnangovindraj committed:
Refactor insert & delete of vertices to resolve kind at runtime

krishnangovindraj committed:
Hide away the inverted label in the Indexer

krishnangovindraj committed:
add compiler/write to checkstyle test

krishnangovindraj committed:
Bump typeql

krishnangovindraj committed:
Missing license header in delete.rs

dmitrii-ubskii committed:
Rename roleplayer edge to links + plan (#7132)

## Usage and product changes

We rename the role player edge to be referred to as a "links" edge, to
avoid confusion with the `RolePlayer` struct which refers to a player in
a relation bundled with the role type tag.

krishnangovindraj committed:
Rustfmt

krishnangovindraj commented:
Merged upstream.
